tests/gen/gen_auto.go

```diff
@@ -0,0 +1,251 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gen
+
+import (
+	"math/rand"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+)
+
+const (
+	// DefaultNumDocs is the default number of documents to generate for a collection.
+	DefaultNumDocs = 10
+	// DefaultNumChildrenPerDoc is the default number of children to generate for a document.
+	DefaultNumChildrenPerDoc = 2
+
+	// DefaultStrLen is the default length of a string to generate.
+	DefaultStrLen = 10
+	// DefaultIntMin is the default minimum value of an integer to generate.
+	DefaultIntMin = 0
+	// DefaultIntMax is the default maximum value of an integer to generate.
+	DefaultIntMax = 10000
+)
+
+type docRec struct {
+	doc    map[string]any
+	docKey string
+}
+
+// AutoGenerateFromSDL generates random documents from a schema.
+func AutoGenerateFromSDL(gqlSDL string, options ...Option) ([]GeneratedDoc, error) {
+	genConfigs, err := parseConfig(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := randomDocGenerator{types: typeDefs, config: genConfigs}
+	return generator.GenerateDocs(options...)
+}
+
+// AutoGenerate generates random documents from collection definitions.
+func AutoGenerate(definitions []client.CollectionDefinition, options ...Option) ([]GeneratedDoc, error) {
+	err := validateDefinitions(definitions)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, def := range definitions {
+		typeDefs[def.Description.Name] = def
+	}
+	generator := randomDocGenerator{types: typeDefs}
+	return generator.GenerateDocs(options...)
+}
+
+type randomDocGenerator struct {
+	types        map[string]client.CollectionDefinition
+	config       configsMap
+	resultDocs   []GeneratedDoc
+	usageCounter typeUsageCounters
+	cols         map[string][]docRec
+	docsDemand   map[string]typeDemand
+	random       rand.Rand
+}
+
+func (g *randomDocGenerator) GenerateDocs(options ...Option) ([]GeneratedDoc, error) {
+	g.cols = make(map[string][]docRec)
+	if g.config == nil {
+		g.config = make(configsMap)
+	}
+
+	configurator := newDocGenConfigurator(g.types, g.config)
+	err := configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *configurator.random
+
+	g.docsDemand = configurator.docsDemand
+	g.usageCounter = configurator.usageCounter
+
+	g.resultDocs = make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	docsLists, err := g.generateRandomDocs(configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, docsList := range docsLists {
+		typeDef := g.types[docsList.ColName]
+		for _, doc := range docsList.Docs {
+			g.resultDocs = append(g.resultDocs, GeneratedDoc{
+				ColName: typeDef.Description.Name,
+				JSON:    createDocJSON(&typeDef, doc),
+			})
+		}
+	}
+	return g.resultDocs, nil
```

Comment:
suggestion: Having now read this function, I cannot really tell why the state-machine like pattern is used here.  I think it would be easier to read/maintain if instead of mutating the host object state you just returned a `results` struct, with whatever stuff needed on it.